File: app/src/main/java/com/ash/studios/musify/Activities/Categories/PlayList.java
package com.ash.studios.musify.Activities.Categories;

import static com.ash.studios.musify.utils.Instance.mp;
import static com.ash.studios.musify.utils.Instance.songs;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.View;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ash.studios.musify.Activities.Player;
import com.ash.studios.musify.Activities.SearchList.CategorySearch;
import com.ash.studios.musify.Adapters.PlaylistAdapter;
import com.ash.studios.musify.Interfaces.IControl;
import com.ash.studios.musify.Interfaces.IService;
import com.ash.studios.musify.Models.Playlist;
import com.ash.studios.musify.Models.Song;
import com.ash.studios.musify.R;
import com.ash.studios.musify.Services.MusicService;
import com.ash.studios.musify.utils.App;
import com.ash.studios.musify.utils.Constants;
import com.ash.studios.musify.utils.Engine;
import com.ash.studios.musify.utils.Instance;
import com.ash.studios.musify.utils.Utils;
import com.bumptech.glide.Glide;
import com.futuremind.recyclerviewfastscroll.FastScroller;
import com.google.android.material.appbar.AppBarLayout;

import java.util.ArrayList;
import java.util.Random;

import me.everything.android.ui.overscroll.OverScrollDecoratorHelper;

public class PlayList extends AppCompatActivity implements
        MediaPlayer.OnCompletionListener, IControl, IService {
    ImageView icon, shufflePlay, sequencePlay, searchBtn, optionsBtn, snippetArt, snippetPlayBtn;
    TextView title, NF, snippetTitle, snippetArtist;
    ConstraintLayout backToLib;
    AppBarLayout appBar;
    ProgressBar loader;
    CardView snippet;
    RecyclerView rv;
    FastScroller fs;

    Engine engine;
    Dialog dialog;
    Context context;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.songs_list);
        getWindow().setStatusBarColor(Color.TRANSPARENT);
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);

        setIDs();
        showAttributes();
        backToLib.setOnClickListener(v -> finish());
        snippet.setOnClickListener(v -> startActivity(new Intent(context, Player.class)));
        new Thread(() -> snippetPlayBtn.setOnClickListener(v -> {
            if (Instance.mp != null) {
                if (Instance.mp.isPlaying()) {
                    snippetPlayBtn.setImageResource(R.drawable.ic_play);
                    Instance.mp.pause();
                    stopService(new Intent(context, MusicService.class));
                } else {
                    snippetPlayBtn.setImageResource(R.drawable.ic_pause);
                    Instance.mp.start();
                    Instance.playing = true;
                    startService(new Intent(context, MusicService.class).setAction(Constants.ACTION.CREATE.getLabel()));
                }
            } else {
                engine.startPlayer();
                snippetPlayBtn.setImageResource(R.drawable.ic_pause);
            }
        })).start();
        optionsBtn.setOnClickListener(v -> {
            dialog = Utils.getDialog(context, R.layout.options_dg);

            TextView dialogName = dialog.findViewById(R.id.dialog_name);
            ImageView dialogIcon = dialog.findViewById(R.id.dialog_icon);
            ConstraintLayout RM = dialog.findViewById(R.id.rescan_media);
            ConstraintLayout LO = dialog.findViewById(R.id.listing_options);
            ConstraintLayout AN = dialog.findViewById(R.id.add_new);
            ConstraintLayout ST = dialog.findViewById(R.id.settings);

            dialogName.setText(title.getText());
            dialogIcon.setImageDrawable(icon.getDrawable());

            ST.setVisibility(View.GONE);
            LO.setVisibility(View.GONE);
            AN.setVisibility(View.VISIBLE);
            RM.setOnClickListener(rm -> {
                dialog.dismiss();

                rv.setAdapter(null);
                NF.setVisibility(View.GONE);
                loader.setVisibility(View.VISIBLE);
                new Handler(Looper.getMainLooper()).postDelayed(() -> {
                    rv.setAdapter(new PlaylistAdapter(context, Utils.getPlaylists(context), loader, NF));
                    if (rv.getAdapter() == null || rv.getAdapter().getItemCount() == 0)
                        disableBtn();
                }, 10);
            });
            AN.setOnClickListener(an -> {
                dialog.dismiss();

                Dialog nameDialog = Utils.getDialog(context, R.layout.name_dg);
                TextView okBtn = nameDialog.findViewById(R.id.ok_btn);
                TextView cancel = nameDialog.findViewById(R.id.cancel_btn);
                EditText plEditText = nameDialog.findViewById(R.id.playlist_edit_text);
                plEditText.requestFocus();

                cancel.setOnClickListener(c -> nameDialog.dismiss());
                okBtn.setOnClickListener(ok -> {
                    String playlistName = plEditText.getText().toString().trim();

                    if (!TextUtils.isEmpty(playlistName)) {
                        Utils.createNewPlaylist(context, playlistName);
                        PlaylistAdapter adapter = new PlaylistAdapter(context, Utils.getPlaylists(context), loader, NF);
                        rv.setAdapter(adapter);
                        adapter.notifyDataSetChanged();
                        nameDialog.dismiss();
                    }

                    if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)
                        showAttributes();
                });
            });
        });
    }

    private void setIDs() {
        context = this;
        engine = new Engine(context);
        rv = findViewById(R.id.song_list);
        appBar = findViewById(R.id.app_bar);
        snippet = findViewById(R.id.snippet);
        NF = findViewById(R.id.nothing_found);
        fs = findViewById(R.id.fast_song_list);
        icon = findViewById(R.id.activity_icon);
        backToLib = findViewById(R.id.lib_back);
        loader = findViewById(R.id.list_loader);
        title = findViewById(R.id.activity_title);
        searchBtn = findViewById(R.id.search_btn);
        optionsBtn = findViewById(R.id.options_btn);
        sequencePlay = findViewById(R.id.play_all_btn);
        shufflePlay = findViewById(R.id.shuffle_all_btn);
        snippetTitle = snippet.findViewById(R.id.snip_title);
        snippetArt = snippet.findViewById(R.id.snip_album_art);
        snippetArtist = snippet.findViewById(R.id.snip_artist);
        snippetPlayBtn = snippet.findViewById(R.id.snip_play_btn);

        snippetTitle.setSelected(true);
        title.setText(new StringBuilder("Playlists"));
        icon.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_playlist));
        icon.setColorFilter(Color.parseColor(getIntent().getStringExtra("icon_color")));

        if (Instance.songs != null) updateSnippet();
        rv.setLayoutManager(new LinearLayoutManager(context));
        rv.setAdapter(new PlaylistAdapter(context, Utils.getPlaylists(context), loader, NF));

        if (rv.getAdapter() == null || rv.getAdapter().getItemCount() == 0) hideAttributes();
        OverScrollDecoratorHelper.setUpOverScroll(rv, OverScrollDecoratorHelper.ORIENTATION_VERTICAL);

        fs.setRecyclerView(rv);
    }

    private void disableBtn() {
        searchBtn.setAlpha(0.4f);
        sequencePlay.setAlpha(0.4f);
        shufflePlay.setAlpha(0.4f);

        NF.setVisibility(View.VISIBLE);
        loader.setVisibility(View.GONE);

        searchBtn.setOnClickListener(null);
        sequencePlay.setOnClickListener(null);
        shufflePlay.setOnClickListener(null);
    }

    private void hideAttributes() {
        shufflePlay.setAlpha(0.4f);
        sequencePlay.setAlpha(0.4f);
        searchBtn.setAlpha(0.4f);

        shufflePlay.setOnClickListener(null);
        sequencePlay.setOnClickListener(null);
        searchBtn.setOnClickListener(null);
    }

    private void updateSnippet() {
        if (Instance.songs.size() > 0) {
            snippet.setVisibility(View.VISIBLE);
            snippetTitle.setText(Instance.songs.get(Instance.position).getTitle());
            snippetArtist.setText(Instance.songs.get(Instance.position).getArtist());
            Glide.with(getApplicationContext())
                    .asBitmap()
                    .placeholder(R.drawable.placeholder)
                    .load(Utils.getAlbumArt(Instance.songs.get(Instance.position).getAlbum_id()))
                    .into(snippetArt);

            int[] accents = Utils.getSecondaryColors(context, Utils.getAlbumArt(Instance.songs.get(Instance.position).getAlbum_id()));
            snippet.setCardBackgroundColor(accents[0]);
            snippetTitle.setTextColor(accents[1]);
            snippetArtist.setTextColor(accents[1]);
            snippetPlayBtn.setColorFilter(accents[1], PorterDuff.Mode.SRC_IN);

            if (Instance.mp != null && Instance.mp.isPlaying())
                snippetPlayBtn.setImageResource(R.drawable.ic_pause);
            else
                snippetPlayBtn.setImageResource(R.drawable.ic_play);
        } else snippet.setVisibility(View.GONE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        ((App) getApplicationContext()).setCurrentContext(context);
        if (Instance.songs != null) updateSnippet();
        if (Instance.mp != null) Instance.mp.setOnCompletionListener(this);
        rv.setAdapter(new PlaylistAdapter(context, Utils.getPlaylists(context), loader, NF));
        if (rv.getAdapter() != null) {
            rv.getAdapter().notifyDataSetChanged();
            if (rv.getAdapter().getItemCount() == 0) hideAttributes();
            else showAttributes();
        }
    }

    private void showAttributes() {
        searchBtn.setAlpha(1f);
        shufflePlay.setAlpha(1f);
        sequencePlay.setAlpha(1f);
        NF.setVisibility(View.GONE);

        searchBtn.setOnClickListener(v -> {
            if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)
                startActivity(new Intent(context, CategorySearch.class)
                        .putExtra("cat_key", 5).putExtra("cat_name", "Playlists"));
        });
        new Thread(() -> {
            if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)

                shufflePlay.setOnClickListener(v -> new Thread(() -> {
                    ArrayList<Song> list = new ArrayList<>();
                    for (Playlist playlist : Utils.getPlaylists(context))
                        list.addAll(playlist.getSongs());

                    shufflePlay.post(() -> {
                        if (list.size() > 0) {
                            Instance.songs = list;
                            Instance.shuffle = true;
                            Instance.position = new Random().nextInt((songs.size() - 1) + 1);

                            engine.startPlayer();
                            mp.setOnCompletionListener(PlayList.this);

                            updateSnippet();
                            Utils.putShflStatus(context, true);
                            Toast.makeText(context, "Shuffle all songs", Toast.LENGTH_SHORT).show();
                        } else
                            Toast.makeText(context, "No songs found in any playlist :(", Toast.LENGTH_SHORT).show();
                    });
                }).start());
        }).start();
        new Thread(() -> {
            if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)

                sequencePlay.setOnClickListener(v -> new Thread(() -> {
                    ArrayList<Song> list = new ArrayList<>();
                    for (Playlist playlist : Utils.getPlaylists(context))
                        list.addAll(playlist.getSongs());

                    sequencePlay.post(() -> {
                        if (list.size() > 0) {
                            Instance.songs = list;
                            Instance.position = 0;
                            Instance.shuffle = false;

                            if (songs.size() > 0) engine.startPlayer();
                            mp.setOnCompletionListener(PlayList.this);

                            updateSnippet();
                            Utils.putShflStatus(context, false);
                            Toast.makeText(context, "Sequence all songs", Toast.LENGTH_SHORT).show();
                        } else
                            Toast.makeText(context, "No songs found in any playlist :(", Toast.LENGTH_SHORT).show();
                    });
                }).start());
        }).start();
    }

    @Override
    public void onStartService() {
        engine.startPlayer();
        updateSnippet();
    }

    @Override
    public void onPrevClicked() {
        engine.playPrevSong();
        updateSnippet();
        startService(new Intent(context, MusicService.class).setAction(Constants.ACTION.CREATE.getLabel()));
    }

    @Override
    public void onNextClicked() {
        engine.playNextSong();
        updateSnippet();
        startService(new Intent(context, MusicService.class).setAction(Constants.ACTION.CREATE.getLabel()));
    }

    @Override
    public void onPlayClicked() {
        if (Instance.mp != null) mp.start();
        else {
            engine.startPlayer();
            mp.setOnCompletionListener(PlayList.this);
        }
        updateSnippet();
    }

    @Override
    public void onPauseClicked() {
        if (Instance.mp != null) Instance.mp.pause();
        updateSnippet();
    }

    @Override
    public void onStopService() {
        if (Instance.mp != null) {
            Instance.mp.stop();
            Instance.mp.release();
            Instance.mp = null;
        }
        snippetPlayBtn.setImageResource(R.drawable.ic_play);
        stopService(new Intent(context, MusicService.class));
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        engine.playNextSong();
        if (Instance.mp != null) {
            Instance.mp = MediaPlayer.create(getApplicationContext(), Instance.uri);
            Instance.mp.start();
            Instance.mp.setOnCompletionListener(this);
        }
        Utils.putCurrentPosition(context, Instance.position);
        updateSnippet();
    }
}

File: app/src/main/java/com/ash/studios/musify/activities/categories/PlayList.kt
package com.ash.studios.musify.activities.categories

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.view.View
import android.view.WindowManager
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ash.studios.musify.adapters.PlaylistAdapter
import com.ash.studios.musify.Interfaces.IControl
import com.ash.studios.musify.Interfaces.IService
import com.ash.studios.musify.Models.Song
import com.ash.studios.musify.R
import com.ash.studios.musify.Services.MusicService
import com.ash.studios.musify.activities.Player
import com.ash.studios.musify.activities.searchList.CategorySearch
import com.ash.studios.musify.utils.App
import com.ash.studios.musify.utils.Constants
import com.ash.studios.musify.utils.Engine
import com.ash.studios.musify.utils.Instance.mp
import com.ash.studios.musify.utils.Instance.playing
import com.ash.studios.musify.utils.Instance.position
import com.ash.studios.musify.utils.Instance.shuffle
import com.ash.studios.musify.utils.Instance.songs
import com.ash.studios.musify.utils.Instance.uri
import com.ash.studios.musify.utils.Utils
import com.bumptech.glide.Glide
import com.futuremind.recyclerviewfastscroll.FastScroller
import com.google.android.material.appbar.AppBarLayout
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper
import java.util.*

class PlayList : AppCompatActivity(), OnCompletionListener, IControl, IService {
    private lateinit var icon: ImageView
    private lateinit var shufflePlay: ImageView
    private lateinit var sequencePlay: ImageView
    private lateinit var searchBtn: ImageView
    private lateinit var optionsBtn: ImageView
    private lateinit var snippetArt: ImageView
    private lateinit var snippetPlayBtn: ImageView
    private lateinit var title: TextView
    private lateinit var notFound: TextView
    private lateinit var snippetTitle: TextView
    private lateinit var snippetArtist: TextView
    private lateinit var backToLib: ConstraintLayout
    private lateinit var appBar: AppBarLayout
    private lateinit var loader: ProgressBar
    private lateinit var snippet: CardView
    private lateinit var rv: RecyclerView
    private lateinit var fs: FastScroller
    private lateinit var engine: Engine
    private lateinit var dialog: Dialog
    private lateinit var context: Context
    @SuppressLint("NotifyDataSetChanged")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.songs_list)
        window.statusBarColor = Color.TRANSPARENT
        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        setIDs()
        showAttributes()
        backToLib.setOnClickListener { finish() }
        snippet.setOnClickListener { startActivity(Intent(context, Player::class.java)) }
        Thread {
            snippetPlayBtn.setOnClickListener {
                if (mp != null) {
                    if (mp!!.isPlaying) {
                        snippetPlayBtn.setImageResource(R.drawable.ic_play)
                        mp!!.pause()
                        stopService(Intent(context, MusicService::class.java))
                    } else {
                        snippetPlayBtn.setImageResource(R.drawable.ic_pause)
                        mp!!.start()
                        playing = true
                        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
                    }
                } else {
                    engine.startPlayer()
                    snippetPlayBtn.setImageResource(R.drawable.ic_pause)
                }
            }
        }.start()
        optionsBtn.setOnClickListener {
            dialog = Utils.getDialog(context, R.layout.options_dg)
            val dialogName = dialog.findViewById<TextView>(R.id.dialog_name)
            val dialogIcon = dialog.findViewById<ImageView>(R.id.dialog_icon)
            val rescanMedia = dialog.findViewById<ConstraintLayout>(R.id.rescan_media)
            val listingOptions = dialog.findViewById<ConstraintLayout>(R.id.listing_options)
            val addNew = dialog.findViewById<ConstraintLayout>(R.id.add_new)
            val settings = dialog.findViewById<ConstraintLayout>(R.id.settings)
            dialogName.text = title.text
            dialogIcon.setImageDrawable(icon.drawable)
            settings.visibility = View.GONE
            listingOptions.visibility = View.GONE
            addNew.visibility = View.VISIBLE
            rescanMedia.setOnClickListener {
                dialog.dismiss()
                rv.adapter = null
                notFound.visibility = View.GONE
                loader.visibility = View.VISIBLE
                Handler(Looper.getMainLooper()).postDelayed({
                    rv.adapter = PlaylistAdapter(context, Utils.getPlaylists(context), loader, notFound)
                    if (rv.adapter == null || rv.adapter!!.itemCount == 0) disableBtn()
                }, 10)
            }
            addNew.setOnClickListener {
                dialog.dismiss()
                val nameDialog = Utils.getDialog(context, R.layout.name_dg)
                val okBtn = nameDialog.findViewById<TextView>(R.id.ok_btn)
                val cancel = nameDialog.findViewById<TextView>(R.id.cancel_btn)
                val plEditText = nameDialog.findViewById<EditText>(R.id.playlist_edit_text)
                plEditText.requestFocus()
                cancel.setOnClickListener { nameDialog.dismiss() }
                okBtn.setOnClickListener {
                    val playlistName = plEditText.text.toString().trim { it <= ' ' }
                    if (!TextUtils.isEmpty(playlistName)) {
                        Utils.createNewPlaylist(context, playlistName)
                        val adapter = PlaylistAdapter(context, Utils.getPlaylists(context), loader, notFound)
                        rv.adapter = adapter
                        adapter.notifyDataSetChanged()
                        nameDialog.dismiss()
                    }
                    if (rv.adapter != null && rv.adapter!!.itemCount > 0) showAttributes()
                }
            }
        }
    }

    private fun setIDs() {
        context = this
        engine = Engine(context)
        rv = findViewById(R.id.song_list)
        appBar = findViewById(R.id.app_bar)
        snippet = findViewById(R.id.snippet)
        notFound = findViewById(R.id.nothing_found)
        fs = findViewById(R.id.fast_song_list)
        icon = findViewById(R.id.activity_icon)
        backToLib = findViewById(R.id.lib_back)
        loader = findViewById(R.id.list_loader)
        title = findViewById(R.id.activity_title)
        searchBtn = findViewById(R.id.search_btn)
        optionsBtn = findViewById(R.id.options_btn)
        sequencePlay = findViewById(R.id.play_all_btn)
        shufflePlay = findViewById(R.id.shuffle_all_btn)
        snippetTitle = snippet.findViewById(R.id.snip_title)
        snippetArt = snippet.findViewById(R.id.snip_album_art)
        snippetArtist = snippet.findViewById(R.id.snip_artist)
        snippetPlayBtn = snippet.findViewById(R.id.snip_play_btn)
        snippetTitle.isSelected = true
        title.text = StringBuilder("Playlists")
        icon.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_playlist))
        icon.setColorFilter(Color.parseColor(intent.getStringExtra("icon_color")))
        if (songs != null) updateSnippet()
        rv.layoutManager = LinearLayoutManager(context)
        rv.adapter = PlaylistAdapter(context, Utils.getPlaylists(context), loader, notFound)
        if (rv.adapter == null || rv.adapter!!.itemCount == 0) hideAttributes()
        OverScrollDecoratorHelper.setUpOverScroll(rv, OverScrollDecoratorHelper.ORIENTATION_VERTICAL)
        fs.setRecyclerView(rv)
    }

    private fun disableBtn() {
        searchBtn.alpha = 0.4f
        sequencePlay.alpha = 0.4f
        shufflePlay.alpha = 0.4f
        notFound.visibility = View.VISIBLE
        loader.visibility = View.GONE
        searchBtn.setOnClickListener(null)
        sequencePlay.setOnClickListener(null)
        shufflePlay.setOnClickListener(null)
    }

    private fun hideAttributes() {
        shufflePlay.alpha = 0.4f
        sequencePlay.alpha = 0.4f
        searchBtn.alpha = 0.4f
        shufflePlay.setOnClickListener(null)
        sequencePlay.setOnClickListener(null)
        searchBtn.setOnClickListener(null)
    }

    private fun updateSnippet() {
        if (songs!!.size > 0) {
            snippet.visibility = View.VISIBLE
            snippetTitle.text = songs!![position].title
            snippetArtist.text = songs!![position].artist
            Glide.with(applicationContext)
                .asBitmap()
                .placeholder(R.drawable.placeholder)
                .load(Utils.getAlbumArt(songs!![position].album_id))
                .into(snippetArt)
            val accents = Utils.getSecondaryColors(context, Utils.getAlbumArt(songs!![position].album_id))
            snippet.setCardBackgroundColor(accents[0])
            snippetTitle.setTextColor(accents[1])
            snippetArtist.setTextColor(accents[1])
            snippetPlayBtn.setColorFilter(accents[1], PorterDuff.Mode.SRC_IN)
            if (mp != null && mp!!.isPlaying) snippetPlayBtn.setImageResource(R.drawable.ic_pause) else snippetPlayBtn.setImageResource(R.drawable.ic_play)
        } else snippet.visibility = View.GONE
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onResume() {
        super.onResume()
        (applicationContext as App).currentContext = context
        if (songs != null) updateSnippet()
        if (mp != null) mp!!.setOnCompletionListener(this)
        rv.adapter = PlaylistAdapter(context, Utils.getPlaylists(context), loader, notFound)
        if (rv.adapter != null) {
            rv.adapter!!.notifyDataSetChanged()
            if (rv.adapter!!.itemCount == 0) hideAttributes() else showAttributes()
        }
    }

    private fun showAttributes() {
        searchBtn.alpha = 1f
        shufflePlay.alpha = 1f
        sequencePlay.alpha = 1f
        notFound.visibility = View.GONE
        searchBtn.setOnClickListener {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) startActivity(
                Intent(context, CategorySearch::class.java)
                    .putExtra("cat_key", 5).putExtra("cat_name", "Playlists")
            )
        }
        Thread {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) shufflePlay.setOnClickListener {
                Thread {
                    val list = ArrayList<Song>()
                    for (playlist in Utils.getPlaylists(context)) list.addAll(playlist.songs)
                    shufflePlay.post {
                        if (list.size > 0) {
                            songs = list
                            shuffle = true
                            position = Random().nextInt(songs!!.size - 1 + 1)
                            engine.startPlayer()
                            mp!!.setOnCompletionListener(this@PlayList)
                            updateSnippet()
                            Utils.putShflStatus(context, true)
                            Toast.makeText(context, "Shuffle all songs", Toast.LENGTH_SHORT).show()
                        } else Toast.makeText(context, "No songs found in any playlist :(", Toast.LENGTH_SHORT).show()
                    }
                }.start()
            }
        }.start()
        Thread {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) sequencePlay.setOnClickListener {
                Thread {
                    val list = ArrayList<Song>()
                    for (playlist in Utils.getPlaylists(context)) list.addAll(playlist.songs)
                    sequencePlay.post {
                        if (list.size > 0) {
                            songs = list
                            position = 0
                            shuffle = false
                            if (songs!!.size > 0) engine.startPlayer()
                            mp!!.setOnCompletionListener(this@PlayList)
                            updateSnippet()
                            Utils.putShflStatus(context, false)
                            Toast.makeText(context, "Sequence all songs", Toast.LENGTH_SHORT).show()
                        } else Toast.makeText(context, "No songs found in any playlist :(", Toast.LENGTH_SHORT).show()
                    }
                }.start()
            }
        }.start()
    }

    override fun onStartService() {
        engine.startPlayer()
        updateSnippet()
    }

    override fun onPrevClicked() {
        engine.playPrevSong()
        updateSnippet()
        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
    }

    override fun onNextClicked() {
        engine.playNextSong()
        updateSnippet()
        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
    }

    override fun onPlayClicked() {
        if (mp != null) mp!!.start() else {
            engine.startPlayer()
            mp!!.setOnCompletionListener(this@PlayList)
        }
        updateSnippet()
    }

    override fun onPauseClicked() {
        if (mp != null) mp!!.pause()
        updateSnippet()
    }

    override fun onStopService() {
        if (mp != null) {
            mp!!.stop()
            mp!!.release()
            mp = null
        }
        snippetPlayBtn.setImageResource(R.drawable.ic_play)
        stopService(Intent(context, MusicService::class.java))
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        engine.playNextSong()
        if (mp != null) {
            mp = MediaPlayer.create(applicationContext, uri)
            mp!!.start()
            mp!!.setOnCompletionListener(this)
        }
        Utils.putCurrentPosition(context, position)
        updateSnippet()
    }
}